fdbserver/VersionedBTree.actor.cpp

```diff
@@ -2871,7 +2871,8 @@ struct RedwoodRecordRef {
 			case 1:
 				return *(int32_t*)r;
 			case 2:
-				return (((int64_t)((int48_t*)r)->high) << 16) | (((int48_t*)r)->low & 0xFFFF);
+				return ((int64_t) static_cast<uint32_t>(reinterpret_cast<const int48_t*>(r)->high) << 16) |
```

Comment:
This one's quite complicated, it might warrant a comment explaining what it's doing.

Reply:
@sfc-gh-satherton I might need a bit of help commenting this accurately. My change was to cast int32_t to uint32_t before left shifting so that we don't have UB if the int32_t is negative.

Reply:
Ok, thanks, I was curious what the error here was and that helps.

I'm noticing that the function returns an `int` but is casting to an `int64_t`, and it probably makes sense to do away with that discrepancy. Also, `int48_t->high` is a 32-bit int, so shifting it 16 bits could theoretically overflow a 32-bit integer if it were large. I'm guessing some precondition of this operation makes that ok, but we should document (and `ASSERT`?) what the contract is here for clarity.

Reply:
And I guess while we're discussing this line, does `(((int48_t*)r)->low & 0xFFFF)` have a different result than `(int48_t*)r->low`?

Reply:
FYI this code is gone in my WIP with a PR coming in within days.  Redwood's versioning has been strictly Pager-level for quite a while, but in case we had any reason to revisit that decision I left some of the optional version handling code in place.  This "version delta" concept in `RedwoodRecordRef` is one example of this.

The BTree always sets `HAS_VERSION` to 0 so version deltas are never used in Redwood instances, however there still is a unit test for `RedwoodRecordRef` which has versions and tests the delta compression code.  

@ajbeamon Returning `int` is just a bug, nice catch.  I guess the unit test did not try large enough version gaps.  [This](https://github.com/apple/foundationdb/blob/8f8bfb1beeaa9189f9d5e8584afacc004a17ce15/fdbserver/VersionedBTree.actor.cpp#L7195-L7200) was my attempt to do so, and looking at it now I have no idea what I was thinking when I wrote it.  I think it would have been sufficient to just use a random 64 bit integer.

I added `& 0xFFFF` because of some real or imagined risk that since the `|` operation is done as `int64_t` when the `int16_t` variable `low` is negative then the upcast will set the 48 upper bits to preserve the upcasted numerical value which would be wrong.  That said, looking at it now I'm not even sure this would fix that problem, I think it would have to be `( (int64_t)((int48_t*)r)->low) & 0xFFFF);`.

Regarding the `<< 16`, the value being shifted is an `int64_t` because of the cast at the start of the statement.  At least that is what I intended.  Adding spaces for clarity, I think it reads as
```
(
  (int64_t) static_cast<uint32_t>(reinterpret_cast<const int48_t*>(r)->high)
   << 16
)
```

Reply:
I think that means that the above line should be casting to a `uint64_t` rather than a `uint32_t` to fix the detected UB (at least once the function is updated to return `int64_t`).

I guess maybe it doesn't matter much, though, if the function is about to go away.

Reply:
I think that's right, but yeah my PR to remove it is imminent.